Modify runtests.py to run a subset of tests based on filenames

terminalmage opened the pull request:
This adds two new options to runtests.py's option parser:

- `--from-filenames` - A comma-separated list of filenames. runtests.py will then attempt to match test modules in both the unit and integration suites, by filename (e.g. `salt/modules/git.py` would result in `unit.modules.test_git` and `integration.modules.test_git` being run.
- `--filename-map` - A supplemental YAML file which can be used to map file globs to test modules, allowing us to make sure that changes in a pkg module cause the pkg/pkgrepo state tests to be run (for example). Each glob will be `fnmatch.filter`'ed against the list of files, and if a match is found, the targeted list of test modules will be run.

Right now, the initial `tests/filename_map.yml` only contains one glob, I need to add more and would like some input on what should be added.

terminalmage committed:
Add --from-filenames and --filename-map options to parser

This allows for the passed filename(s) to decide which tests are run

terminalmage committed:
Add an initial filename_map

terminalmage committed:
Don't run unit or integration tests unnecessarily

If --from-filenames (and optionally --filename-map) yield no matches,
this will keep from running the entire unit/integration suite.

cachedout commented:
This is cool. I really like this approach.

terminalmage committed:
Enhancements to --from-filenames, --filename-map options

This commit adds the following features:

- Changes to test modules are identified, and those tests are added to
  the list of tests to be run
- expr_match is used to allow the filename map to use either globbing or
  regular expressions
- Corresponding state tests are added when doing filename matches,
  meaning that for example the git state tests will be run when the git
  execution module is in the --from-filenames list

terminalmage committed:
Fill out the filename map a bit

terminalmage commented:
I've added some functionality to `tests/support/parsers/__init__.py` (details in [this commit](https://github.com/saltstack/salt/pull/47337/commits/09ac5114103d144bb4afb228a3f51b8a946424dc). I've also filled out the filename map with what I could identify with a couple passes through the state/execution modules.

terminalmage committed:
Add more unit test mapping

rallytime commented:
@terminalmage This has a lint error in it: https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/21741/violations/file/tests/support/parser/__init__.py/

Can you fix that up?

rallytime reviewed (commented):
This is great! I just have one small question, plus the lint error should be fixed. Otherwise, really nice.

rallytime commented on tests/support/parser/__init__.py:
```diff
@@ -33,7 +35,19 @@
+
+try:
+    # 2018.3 and later
+    from salt.utils.stringutils import expr_match as _expr_match
```
@terminalmage Right, but you have files above that are not in a try/except block that were also moved around during the `salt.utils` changes. My thought is that either all of those imports should be moved into a try/except, or none should.

terminalmage replied:
Ahh yeah, I added those other imports later after initially doing it that way for expr_match.  I removed the try/except. We'll either have to backport separately for 2017.7 and 2018.3 or not backport past 2018.3.

terminalmage committed:
Move helper outside of for loop

terminalmage committed:
Don't import old module name

We'll handle the old name when backporting

rallytime commented:
@terminalmage Bump - did you see the feedback from @Ch3LL?

terminalmage committed:
Move context cache tests to their proper location

terminalmage committed:
Move salt.utils.context tests to their proper location

terminalmage committed:
Use tests.support.paths instead of deprecated integration module

terminalmage committed:
Rename test pgjsonb returner test file to match naming convention

terminalmage committed:
Rename sentry returner test to match naming convention

terminalmage committed:
Fix bad copypasta in docstring

terminalmage committed:
Move jinja templating tests alongside salt/utils/jinja.py's tests

terminalmage committed:
Remove redundant "test" from filename

terminalmage committed:
Rename the aptpkg state module to respect naming convention

terminalmage committed:
Rename test_map_conf test file to respect naming convention

terminalmage committed:
Rename slack engine test file to respect naming convention

terminalmage committed:
Consolidate misplaced tests into tests/unit/test_config.py

terminalmage committed:
Move shared kernelpkg test cases to tests/support/

terminalmage committed:
Move inspectlib tests to conform with naming convention

terminalmage committed:
Rename beacon unit test modules to conform with naming convention

Also consolidate pyroute2 network_settings tests into network_settings
beacon test module.

terminalmage committed:
Consolidate and move loader tests to test/unit/test_loader.py

terminalmage committed:
Consolitdate and move salt-ssh unit tests to proper location

terminalmage committed:
Consolidate netapi tests and move them to reflect naming convention

terminalmage committed:
Consolidate saltclass pillar unit tests into single module

terminalmage committed:
Rename pillar unit test modules to reflect naming convention

terminalmage committed:
Rename/consolidate salt.utils unit tests to conform to naming convention

terminalmage committed:
Consolidate/move fileclient/fileserver tests to conform to naming convention

terminalmage committed:
Rename acl unit test module to confirm to naming convention

terminalmage committed:
Rename custom_grains.py so it's not identified as a test module

terminalmage committed:
Rename some files to avoid them being identified as test modules

terminalmage committed:
Split mysql database tests into two modules to respect naming convention

terminalmage committed:
Split ssh state tests into separate modules to respect naming convention

terminalmage committed:
Rename virtualenv state test module to respect naming convention

terminalmage committed:
Rename "providers" dir to "clouds" to respect naming convention

terminalmage committed:
Rename providers to clouds to respect naming convention

Also, move rackspace openstack-based tests to test_openstack.py

terminalmage committed:
Rename exec module integration test files to respect naming convention

terminalmage committed:
Rename venafi test module to bring it closer to the naming convention

terminalmage committed:
Add function to get all of the test modules

terminalmage committed:
Use the new helper to get the test mods

terminalmage committed:
Add additional paths to the filename map

terminalmage committed:
Add back test accidentally deleted when cherry-picking

terminalmage committed:
Add back changes lost in cherry-pick

terminalmage committed:
Rename files to match naming convention

terminalmage committed:
Add integration.netapi.rest_tornado.test_app to filename map

terminalmage committed:
Add test to check test module names and make sure they match the naming convention

terminalmage committed:
Make sure we always run the module names test

terminalmage commented:
@cachedout Perhaps?

terminalmage commented:
This PR has been updated with the commits from https://github.com/saltstack/salt/pull/49091, as well as a unit test to enforce the naming convention. It still needs some local testing before it's ready for re-review, however.

terminalmage committed:
Lint

terminalmage committed:
Moar Lint

terminalmage committed:
Evan Moar Lint

terminalmage commented on tests/filename_map.yml:
```diff
@@ -0,0 +1,96 @@
+
+salt/modules/(mac_user|useradd|pw_user|solaris_user|win_useradd)\.py:
+  - unit.states.test_user
+  - integration.states.test_user
```
Yes, this is all on purpose. runtests.py will automagically match files that match a naming convention. Otherwise, the map file would be thousands of lines long and would be a pain in the butt to maintain.

So, there is no need to add `integration.modules.test_useradd` and `unit.modules.test_useradd` as they would be picked up when `salt/modules/useradd.py` is specified in the `--from-filenames` parameter.

terminalmage committed:
Fix conflicting dict key in filename map

terminalmage committed:
Fix import for moved module

rallytime reviewed (approved):
Well done! I am very excited about this.

Pending tests, 👍

terminalmage commented:
There are failing unit tests for the loader which were fixed in every branch but develop, so I opened https://github.com/saltstack/salt/pull/49127 to merge fluorine into develop. I confirmed that the failing tests failed locally when run against develop and then passed once I merged forward.

terminalmage commented:
This PR should be rebased once https://github.com/saltstack/salt/pull/49127 is merged into develop.

cachedout commented:
@terminalmage We're ready for that rebase now. :)

cachedout commented on tests/filename_map.yml:
```diff
@@ -0,0 +1,229 @@
+'*':
```
So, I presume that the workflow for adding tests would change so that this file should be modified when appropriate when new tests are added? If so, have we documented that in the documentation on writing Salt tests?

terminalmage commented:
Rebased.

gtmanfred commented:
@terminalmage there are a couple failed tests, they don't look like they would be related to me, but would you mind double checking?

terminalmage commented:
@gtmanfred Already ahead of you. The pkg test is unrelated and I've opened https://github.com/saltstack/salt/pull/49134 to fix it. Looking at the sdb test now.

terminalmage commented:
sdb tests are failing on the develop branch already and are also unrelated.
